eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs

```diff
@@ -12,44 +13,57 @@
 module Cardano.Ledger.Conway.Rules.VDel (
   ConwayVDEL,
   ConwayVDelEvent (..),
-  VDelEnv (..),
   ConwayVDelPredFailure,
 )
 where
 
 import Cardano.Ledger.BaseTypes (
   ShelleyBase,
  )
-import Cardano.Ledger.CertState (VState)
+import Cardano.Ledger.CertState (VState (VState, vsCCHotKeys, vsDReps))
 import Cardano.Ledger.Conway.Era (ConwayVDEL)
-import Cardano.Ledger.Conway.TxCert (ConwayCommitteeCert)
-import Cardano.Ledger.Core (Era (EraCrypto), EraRule)
+import Cardano.Ledger.Conway.TxCert (ConwayCommitteeCert (..))
+import Cardano.Ledger.Core (Era (EraCrypto), EraPParams, EraRule, PParams)
+import Cardano.Ledger.Credential (Credential)
+import Cardano.Ledger.Keys (KeyHash, KeyRole (CommitteeColdKey, Voting))
 import Control.DeepSeq (NFData)
-import Control.State.Transition (
+import Control.State.Transition.Extended (
   BaseM,
   Environment,
   Event,
   PredicateFailure,
   STS,
   Signal,
   State,
+  TRC (TRC),
+  TransitionRule,
+  judgmentContext,
   transitionRules,
+  (?!),
  )
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
 import GHC.Generics (Generic)
 import NoThunks.Class (NoThunks (..))
 
 data ConwayVDelPredFailure era
-  deriving (Show, Eq, Generic, NoThunks, NFData)
+  = ConwayDRepAlreadyRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayDRepNotRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayCommitteeAlreadyRegisteredVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
```

Comment:
This isn't necessary. The point of the hot/cold key scheme is to be able to change the hot key, so being already registered cannot be a failure.

The error message right below also isn't right. The only way registering can fail is if that CC member has resigned. So one error named something like `ConwayCommitteeHasResignedVDEL` would be appropriate.